sklearn/decomposition/online_lda.py

```diff
@@ -0,0 +1,668 @@
+    if (X < 0).any():
+        raise ValueError("Negative values in data passed to %s" % whom)
+
+
+def _dirichlet_expectation(X):
+    """
+    For an array theta ~ Dir(X), computes `E[log(theta)]` given X.
+
+    Parameters
+    ----------
+    X : array-like
+        1 or 2 dimention vector
+
+    Returns
+    -------
+    dirichlet_expect: array-like
+        Dirichlet expectation of input array X
+    """
+
+    if len(X.shape) == 1:
+        dirichlet_expect = _dirichlet_expectation_1d(X)
+    else:
+        dirichlet_expect = _dirichlet_expectation_2d(X)
+    return dirichlet_expect
+
+
+def _update_doc_distribution(X, exp_topic_word_distr, doc_topic_prior, max_iters,
+                             mean_change_tol, cal_diff):
+    """
+    E-step: update document topic distribution.
+    (In the literature, this is called `gamma`.)
+
+    Parameters
+    ----------
+    X : sparse matrix, shape=(n_samples, n_features)
+        Document word matrix.
+
+    exp_topic_word_distr : dense matrrix, shape=(n_topics, n_features)
+        Exponential value of expection of log topic word distribution.
+        In the literature, this is `exp(E[log(beta)])`.
+
+    doc_topic_prior : float
+        Prior of document topic distribution `theta`.
+
+    max_iters : int
+        Max number of iterations for updating document topic distribution in E-step.
+
+    mean_change_tol : float
+        Stopping tolerance for updating document topic distribution in E-setp.
+
+    cal_diff : boolean
+        Parameter that indicate to calculate differene in `component_` or not.
+        Set `cal_diff` to `True` when we need to run M-step.
+
+    Returns
+    -------
+    (doc_topic_distr, component_diff) :
+        `doc_topic_distr` is unnormailzed topic distribution for each document.
+        In the literature, this is `gamma`. we can calcuate `E[log(theta)]`
+        from it.
+        `component_diff` is the difference of `component_` when `cal_diff` is True.
+        Otherwise, it is None.
+
+    """
+
+    n_samples, n_features = X.shape
+    n_topics = exp_topic_word_distr.shape[0]
+
+    doc_topic_distr = np.ones((n_samples, n_topics))
+    # In the literature, this is `exp(E[log(theta)])`
+    exp_doc_topic = np.exp(_dirichlet_expectation(doc_topic_distr))
+
+    # diff on `component_` (only calculate it when `cal_diff` is True)
+    component_diff = np.zeros(exp_topic_word_distr.shape) if cal_diff else None
+
+    X_data = X.data
+    X_indices = X.indices
+    X_indptr = X.indptr
+
+    for d in xrange(n_samples):
+        ids = X_indices[X_indptr[d]:X_indptr[d + 1]]
+        cnts = X_data[X_indptr[d]:X_indptr[d + 1]]
+
+        doc_topic_d = doc_topic_distr[d, :]
+        exp_doc_topic_d = exp_doc_topic[d, :]
+        exp_topic_word_d = exp_topic_word_distr[:, ids]
+
+        # The optimal phi_{dwk} is proportional to 
+        # exp(E[log(theta_{dk})]) * exp(E[log(beta_{dw})]).
+        norm_phi = np.dot(exp_doc_topic_d, exp_topic_word_d) + 1e-100
```

Comment:
To my understanding, Hoffman just don’t want to store the `phi` vector.
And since  `phi_{dwk} proportional to expElog( theta_{dk}) * expElog( theta_{dk})`,
we can rewrite `sum( phi_{dwk} * n_{dw} )`
to `sum( (expElog( theta_{dk}) * expElog( theta_{dk}) / phi_norm) * n_{dw})`
which leads to the `gamma_{dk}` update in his original code [here]
(https://github.com/qpleple/online-lda-vb/blob/master/mdhoffma/onlineldavb.py#L179-L180).

And yes, the reference is the NMF paper. Maybe he is inspired by the updating rules sections. (Section 4 & 5)